go/client/logclient.go

```diff
@@ -109,168 +104,37 @@ type GetProofByHashResponse struct {
-	}
-
-	verifier, err := ct.NewSignatureVerifier(pubkey)
-	if err != nil {
-		return nil, err
-	}
-
-	if hc == nil {
-		hc = new(http.Client)
-	}
-	return &LogClient{uri: uri, httpClient: hc, verifier: verifier}, nil
-}
-
-// Makes a HTTP call to |uri|, and attempts to parse the response as a
-// JSON representation of the structure in |res|. Uses |ctx| to
-// control the HTTP call (so it can have a timeout or be cancelled by
-// the caller), and |httpClient| to make the actual HTTP call.
-// Returns a non-nil |error| if there was a problem.
-func fetchAndParse(ctx context.Context, httpClient *http.Client, uri string, res interface{}) error {
-	req, err := http.NewRequest(http.MethodGet, uri, nil)
-	if err != nil {
-		return err
-	}
-	req.Cancel = ctx.Done()
-	resp, err := httpClient.Do(req)
-	if err != nil {
-		return err
-	}
-	defer resp.Body.Close()
-	// Make sure everything is read, so http.Client can reuse the connection.
-	defer ioutil.ReadAll(resp.Body)
-
-	if resp.StatusCode != 200 {
-		return fmt.Errorf("got HTTP Status %s", resp.Status)
-	}
-
-	if err := json.NewDecoder(resp.Body).Decode(res); err != nil {
-		return err
-	}
-
-	return nil
-}
-
-// Makes a HTTP POST call to |uri|, and attempts to parse the response as a JSON
-// representation of the structure in |res|.
-// Returns a non-nil |error| if there was a problem.
-func (c *LogClient) postAndParse(uri string, req interface{}, res interface{}) (*http.Response, string, error) {
-	postBody, err := json.Marshal(req)
-	if err != nil {
-		return nil, "", err
-	}
-	httpReq, err := http.NewRequest(http.MethodPost, uri, bytes.NewReader(postBody))
-	if err != nil {
-		return nil, "", err
-	}
-	httpReq.Header.Set("Content-Type", "application/json")
-	resp, err := c.httpClient.Do(httpReq)
-	// Read all of the body, if there is one, so that the http.Client can do
-	// Keep-Alive:
-	var body []byte
-	if resp != nil {
-		body, err = ioutil.ReadAll(resp.Body)
-		resp.Body.Close()
-	}
-	if err != nil {
-		return resp, string(body), err
-	}
-	if resp.StatusCode == 200 {
-		if err != nil {
-			return resp, string(body), err
-		}
-		if err = json.Unmarshal(body, &res); err != nil {
-			return resp, string(body), err
-		}
-	}
-	return resp, string(body), nil
-}
-
-func backoffForRetry(ctx context.Context, d time.Duration) error {
-	backoffTimer := time.NewTimer(d)
-	if ctx != nil {
-		select {
-		case <-ctx.Done():
-			return ctx.Err()
-		case <-backoffTimer.C:
-		}
-	} else {
-		<-backoffTimer.C
-	}
-	return nil
+	logClient, err := jsonclient.New(uri, hc, pemEncodedKey)
+	return &LogClient{*logClient}, err
 }
 
 // Attempts to add |chain| to the log, using the api end-point specified by
 // |path|. If provided context expires before submission is complete an
 // error will be returned.
 func (c *LogClient) addChainWithRetry(ctx context.Context, ctype ct.LogEntryType, path string, chain []ct.ASN1Cert) (*ct.SignedCertificateTimestamp, error) {
+	if ctx == nil {
+		ctx = context.TODO()
```

Comment:
The focus here was more on extracting reusable bits of code for v2 use than on fixing v1 client use; added issue #1326.